RPI5: Add SCMI pinctrl, agents and SCMI devices support

GrygiriiS opened the pull request:
This PR  :

- adds generic pinctrl interface
- adds bcm2712 pinctrl driver
- adds scmi pinctrl protocol
- implements calls for supporting multiple SCMI agents and configuring device accesses for them.

This is incorporates the PR https://github.com/xen-troops/arm-trusted-firmware/pull/16 from @lorc

Commit:
pinctrl: Add pinctrl generic interface

This defines the generic interface, and ops, which should provide
the platform specific implementation of pin control subsystem.
This driver was designed to be used from agents via SCP, as well as from
the other drivers, which requires pins configuration.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
pinctrl: bcm2712: add pinctrl driver

Add RPI5 BCM2712 pinctrl driver.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
scmi: implement BASE_DISCOVER_AGENT call

This call provides basic information about an agent.

Change-Id: I857ca7f09bd85679e4743e81cd0762d865e6072a
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
scmi: implement access control calls

Implement BASE_SET_DEVICE_PERMISSIONS and
BASE_RESET_AGENT_CONFIGURATION calls. Add platform-specific callback
that return number of SCMI devices, array with device permission and a
function to reset a SCMI device to a default state.

Change-Id: Idd0e17b95649102c5b8a46c8fb41cbee730dc7d6
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: scmi: add pin control protocol

Add ARM SCMI Pin control protocol implementation according to the SCMI spec
v3.2.

Limitation:
- PINCTRL_SETTINGS_GET supports only one Config per request.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
plat: rpi: rpi5: enable scmi Pin control protocol

Enable ARM SCMI Pin control protocol for RPI5.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
rpi5: scmi: add list of SCMI devices

Provide list of all SCMI devices with their ID's. Provide callbacks
that:
- return device count;
- sets device permissions;
- resets agent configuration.

The RPI5 SCMI platform device and agent management implementation is simple
skeleton code providing minimal features and can be extended in the future
for more complex use-cases.

Change-Id: Ic0b5ee8b4c03be29b31fb76b1740cc4f329a8537
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
scmi: extend multiple agent support

Existing SCMI code already has parts for support multiple agents. Extend
this by providing:

- plat_scmi_agent_count() function that returns the number of agents
supported by a platform.
- plat_scmi_agent_get_name() function that returns agent name.

And ensure proper agent_count passed to the agents in
SCMI_PROTOCOL_ATTRIBUTES message.

Change-Id: I60374842f6fbf457c5da6459161ead301920cfd2
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
plat: rpi5: scmi: extend multiple agent support

Define 8 agents/channels for the RPI5 platform and required SCMI callbacks.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

GrygiriiS commented:
@oleksiimoisieiev @lorc

oleksiimoisieiev commented:
Got compilation error:
```
plat/rpi/rpi5/scmi/scmi.c:14:10: fatal error: device.h: No such file or directory
   14 | #include "device.h"
      |          ^~~~~~~~~~
compilation terminated.
make: *** [Makefile:1513: /home/amoiseev/work/rpi5/arm-trusted-firmware/build/rpi5/debug/bl31/scmi.o] Error 1
make: *** Waiting for unfinished jobs....
  CC      services/arm_arch_svc/arm_arch_svc_setup.c 
cp: cannot stat './build/rpi5/debug/bl31.bin': No such file or directory
```

Commit:
drivers: scmi-msg: reset_domain: add permission check plat callback

Add platform plat_scmi_rstd_permitted() callback to check SCMI Agent access
permissions for Reset domain.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
plat: rpi5: scmi: add device resources management policy

Add RPI5 device resources management policy. Each device is described by
struct scmi_device which allows to define resets and pinctrl resources used
by device (resources). All Device's resources allowed to be accessed by
Agent when the Agent is given access to the Device.

The access to the Device resource expected to be denied if Agent is
not given permission to access this resource.

The current patch implements permission checking only for SCMI reset
domains for now.

NOTE. Shared access to the Device resources is expected to be managed by
corresponding platform drivers (including, but not limited to, reference
counters).

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
plat: rpi5: scmi_policy: update to handle pinctrl

Update RPI5 SCMI policy to handle pinctrl.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

GrygiriiS commented:
- re-based on top of pinctrl
- perms tested w/wo xen

oleksiimoisieiev commented:
tested with resets on dom0 and domd with pcie1 and pcie2
`Tested-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

oleksiimoisieiev commented:
LGTM `Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

lorc commented on include/drivers/pinctrl.h:
```diff
@@ -0,0 +1,211 @@
+
+/*
+ * The following inlines stuffs a configuration parameter and data value
+ * into and out of an unsigned long argument, as used by the generic pin config
```
I believe you wanted to say "unsigned int", not "unsigned long"

GrygiriiS replied:
fixed

lorc commented on include/drivers/pinctrl.h:
```diff
@@ -0,0 +1,211 @@
+
+struct pinctrl_pin {
+	uint32_t pin;
+	uint32_t configs;
```
why "configs" in plural? Should it be "config"?

GrygiriiS replied:
removed

lorc commented on include/drivers/pinctrl.h:
```diff
@@ -0,0 +1,211 @@
+				uint32_t *num_groups);
+
+/* pinmux functions */
+int pinctrl_request_group(uint32_t group_id);
```
What are the purpose for the next 4 functions? As I can see, you do not implement them in BRCM2712 driver.

GrygiriiS replied:
It's called by Linux agent by using PINCTRL_REQUEST. The BRCM2712 has nothing to do, so it implements stubs which always succeed.
More over, for reason unknown to me, Linux agent accessing both pin_groups and pins, and expect both of them to be processed without failure. Therefor  BRCM2712 is implemented considered group_id  == pin_id (the same way as Linux does).

GrygiriiS replied:
UPD: The SCMI spec says - "This command is used by an agent to request exclusive control of a pin or group."
But we actually solve it through Device management - so for RPI5 it's "always ok" stubs.

lorc commented on drivers/scmi-msg/scmi_pinctrl.c:
```diff
@@ -0,0 +1,890 @@
+
+#include "common.h"
+
+#pragma weak pinctrl_get_groups_count
```
I believe ARM TF coding style requires those pragmas to be near the function definition

GrygiriiS replied:
It follows the same style as any other scmi-msg files.

GrygiriiS replied:
UPD. I've checked - seems all weaks in ARM TF defined this way.

lorc commented on drivers/scmi-msg/scmi_pinctrl.c:
```diff
@@ -0,0 +1,890 @@
+#include "common.h"
+
+#pragma weak pinctrl_get_groups_count
+#pragma weak pinctrl_get_functions_count
```
Also, why are you defining default pinctrl function implementation inside the SCMI code? This is the wrong place.

GrygiriiS replied:
I follow the same style as any other scmi-msg files with assumption that they should be able to built without any other dependencies. Is this wrong?

GrygiriiS replied:
Actually I can rework it anyway you wish, but what exactly should i do?

lorc replied:
Well, this is very confusing to have pinctrl interface stubs in SCMI code. 

As far as I understand, we should not have this stubs at all: if a platform does not have pinctrl driver it should not provide pinctrl interface.

But if you want to make stubs anyways, the should live somewhere in `drivers/pinctrl/`, because everyone will except to find them here.

GrygiriiS replied:
fixed

lorc commented on drivers/scmi-msg/scmi_pinctrl.c:
```diff
@@ -0,0 +1,890 @@
+	VERBOSE("scmi: pinctrl attr id = %d flags = %d\n",
+		in_args->identifier, in_args->flags);
+
+	selector = in_args->identifier;
```
You need to use `SPECULATION_SAFE_VALUE` here

GrygiriiS replied:
Honestly I'm not fully understand why? Input msg is stored in local-per-cpu buffer and it's not a shmem.

lorc commented on drivers/scmi-msg/scmi_pinctrl.c:
```diff
@@ -0,0 +1,890 @@
+
+	return_values.status = SCMI_SUCCESS;
+
+	switch (in_args->flags) {
```
`SPECULATION_SAFE_VALUE`

lorc commented on drivers/scmi-msg/scmi_pinctrl.c:
```diff
@@ -0,0 +1,890 @@
+	VERBOSE("scmi: pinctrl assoc group sel = %d skip = %d\n",
+		tx->identifier, tx->index);
+
+	ret = pinctrl_get_group_pins(tx->identifier, &pins, &nr_pins);
```
`SPECULATION_SAFE_VALUE`

lorc commented on drivers/scmi-msg/scmi_pinctrl.c:
```diff
@@ -0,0 +1,890 @@
+		}
+
+		rx->array[lcounter] = pins[counter];
+		rx_size += sizeof(uint16_t);
```
nit: `sizeof(rx->array[0])` would be better

GrygiriiS replied:
fixed

lorc commented on drivers/scmi-msg/scmi_pinctrl.h:
```diff
@@ -0,0 +1,197 @@
+#define SCMI_PINCTRL_ATTR_NAME_SZ 16U
+#define SCMI_PINCTRL_NAME_MAX_SZ  64U
+
+#define ALIGN_NEXT(VALUE, INTERVAL)                                            \
```
There is already `round_up()` from `utils_def.h`, no need for own implementation

GrygiriiS replied:
fixed

lorc commented on drivers/scmi-msg/scmi_pinctrl.h:
```diff
@@ -0,0 +1,197 @@
+	SCMI_PINCTRL_SET_PERMISSIONS = 0xa,
+};
+
+#define FLD(mask, val)     (((val) << (__builtin_ffsll(mask) - 1) & (mask)))
```
Are you sure that this builtin is available in all supported compilers? There is `GENMASK_` macros in `utils_def.h`

GrygiriiS replied:
reworked/removed

lorc commented on plat/rpi/rpi5/rpi5_setup.c:
```diff
@@ -16,7 +16,7 @@ void plat_rpi_bl31_custom_setup(void)
 	generic_delay_timer_init();
 
-#ifdef SCMI_SERVER_SUPPORT
+#if SCMI_SERVER_SUPPORT
```
Stray change?

GrygiriiS replied:
fixed

lorc commented on plat/rpi/rpi5/scmi/scmi.c:
```diff
@@ -3,13 +3,14 @@
-#include <assert.h>
-#include <stdint.h>
 
+#include <assert.h>
```
Stray change

GrygiriiS replied:
fixed

lorc commented on plat/rpi/rpi5/scmi/scmi_policy.c:
```diff
@@ -0,0 +1,80 @@
+
+static uint32_t rpi_scmi_device_owner[RPI5_SCMI_DEV_COUNT];
+
+int32_t plat_scmi_device_permission(uint32_t agent_id, uint32_t device_id,
```
Why you moved this code from `base.c` ? There is nothing RPI5- specific in this functions.

GrygiriiS replied:
Everything here is RPI5 specific actually. Why? because comparing to the bogus RPI5 platform - the normal HW platforms will do much more things here, like configuring firewalls, security settings and etc. And not just saving an Device owner.

lorc commented on plat/rpi/rpi5/scmi/scmi_policy.c:
```diff
@@ -12,18 +12,77 @@
+		.pins = (int[]){-1},
+};
+
+static const struct scmi_device rpi_scmi_devices[RPI5_SCMI_DEV_COUNT] = {
```
This is quite complex logic and I see no benefit in it. Instead of mapping devices to resets, you can map resets to devices. E.g. create simple array "reset_to_dev = [0, RPI5_SCMI_DEV_PCIE1, RPI5_SCMI_DEV_PCIE2, 0 , 0], etc. It will be much easier any simpler.

GrygiriiS replied:
First, this logic follows approach:
- already used by xen-troops in other projects;
- which follows DT definitions.

Second, I can't agree that it will be simpler - lets imagine we have not 4, but 64 resets. It will be total pain trying to fill device table with approach you've proposed. With current approach - all that is required is to follow DT.

GrygiriiS reviewed (commented):
I commented where I have questions. Other comments I'll try to fix.

lorc commented on drivers/scmi-msg/base.c:
```diff
@@ -166,6 +166,46 @@ static void discover_list_protocols(struct scmi_msg *msg)
+		agent_id = msg->agent_id;
+	}
+
+	return_values.name[0] = '\0';
```
I can see that you changed what name is returned in the code, but not in the commit message. Please align one with other.

lorc reviewed (commented):
Please address my last comment and with that:

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

I believe this code is good enough for our use case, but probably there will be additional comments from maintainers during upstream.
